<केंद्र><h1>अध्याय 3 सूचकांक</h1></केंद्र>

In [1]:
import numpy as np
import pandas as pd

## 1. अनुक्रमणिका
### 1. टेबल कॉलम इंडेक्स
कॉलम इंडेक्स सबसे सामान्य इंडेक्स फॉर्म है, जिसे आम तौर पर `[]` द्वारा कार्यान्वित किया जाता है। `[कॉलम नाम]` का उपयोग `डेटाफ़्रेम` से संबंधित कॉलम को निकालने के लिए किया जा सकता है, और रिटर्न मान `सीरीज़` है। उदाहरण के लिए, निकालें तालिका से नाम स्तंभ:

In [2]:
df = pd.read_csv('../data/learn_pandas.csv', usecols = ['School', 'Grade', 'Name', 'Gender', 'Weight', 'Transfer'])
df['Name'].head()

0      Gaopeng Yang
1    Changqiang You
2           Mei Sun
3      Xiaojuan Sun
4       Gaojuan You
Name: Name, dtype: object

यदि आप एकाधिक कॉलम निकालना चाहते हैं, तो आप `[कॉलम नामों की सूची]` का उपयोग कर सकते हैं, और रिटर्न मान एक `डेटाफ़्रेम` है। उदाहरण के लिए, तालिका से लिंग और नाम कॉलम निकालें:

In [3]:
df[['Gender', 'Name']].head()

Gender            Name
0  Female    Gaopeng Yang
1    Male  Changqiang You
2    Male         Mei Sun
3  Female    Xiaojuan Sun
4    Male     Gaojuan You

इसके अलावा, यदि आप एक कॉलम निकालना चाहते हैं और कॉलम नाम में रिक्त स्थान नहीं है, तो आप इसे निकालने के लिए `.column name` का उपयोग कर सकते हैं, जो `[कॉलम नाम]` के बराबर है:

In [4]:
df.Name.head()

0      Gaopeng Yang
1    Changqiang You
2           Mei Sun
3      Xiaojuan Sun
4       Gaojuan You
Name: Name, dtype: object

### 2. अनुक्रम का पंक्ति सूचकांक

[ए] स्ट्रिंग इंडेक्स के साथ `श्रृंखला`

यदि आप किसी एकल सूचकांक के संगत तत्व को निकालना चाहते हैं, तो आप `[आइटम]` का उपयोग कर सकते हैं। यदि `श्रृंखला` में केवल एक ही संगत मान है, तो स्केलर मान लौटाया जाता है। यदि कई संगत मान हैं, तो एक `श्रृंखला ` लौटाया गया है:

In [5]:
s = pd.Series([1, 2, 3, 4, 5, 6], index=['a', 'b', 'a', 'a', 'a', 'c'])
s['a']

a    1
a    3
a    4
a    5
dtype: int64

In [6]:
s['b']

2

यदि आप एकाधिक सूचकांकों के संबंधित तत्वों को पुनः प्राप्त करना चाहते हैं, तो आप `[वस्तुओं की सूची]` का उपयोग कर सकते हैं:

In [7]:
s[['c', 'b']]

c    6
b    2
dtype: int64

यदि आप दो इंडेक्स के बीच के तत्वों को निकालना चाहते हैं, और ये दोनों इंडेक्स पूरे इंडेक्स में अद्वितीय हैं, तो आप एक स्लाइस का उपयोग कर सकते हैं, और ध्यान दें कि यहां स्लाइस में दो एंडपॉइंट होंगे:

In [8]:
s['c': 'b': -2]

c    6
a    4
b    2
dtype: int64

यदि आगे और पीछे के समापन बिंदुओं के मान दोहराए जाते हैं, तो स्लाइस का उपयोग करने से पहले उन्हें क्रमबद्ध करने की आवश्यकता होती है:

In [9]:
try:
    s['a': 'b']
except Exception as e:
    Err_Msg = e
Err_Msg

KeyError("Cannot get left slice bound for non-unique label: 'a'")

In [10]:
s.sort_index()['a': 'b']

a    1
a    3
a    4
a    5
b    2
dtype: int64

[बी] पूर्णांक सूचकांक के साथ `श्रृंखला`

डेटा रीडिंग फ़ंक्शन का उपयोग करते समय, यदि संबंधित कॉलम को इंडेक्स के रूप में निर्दिष्ट नहीं किया जाता है, तो 0 से शुरू होने वाला एक पूर्णांक इंडेक्स डिफ़ॉल्ट इंडेक्स के रूप में उत्पन्न होगा, लंबाई आवश्यकताओं को पूरा करने वाले पूर्णांकों का कोई भी सेट इंडेक्स के रूप में उपयोग किया जा सकता है .

स्ट्रिंग्स की तरह, यदि `[int]` या `[int_list]` का उपयोग किया जाता है, तो संबंधित सूचकांक **तत्व** का मान पुनर्प्राप्त किया जा सकता है:

In [11]:
s = pd.Series(['a', 'b', 'c', 'd', 'e', 'f'], index=[1, 3, 1, 2, 5, 4])
s[1]

1    a
1    c
dtype: object

In [12]:
s[[2,3]]

2    d
3    b
dtype: object

यदि आप पूर्णांक स्लाइसिंग का उपयोग करते हैं, तो संबंधित सूचकांक **स्थिति** का मान निकाल लिया जाएगा। ध्यान दें कि यहां पूर्णांक स्लाइसिंग में सही समापन बिंदु शामिल नहीं है, ठीक `पायथन` में स्लाइसिंग की तरह:

In [13]:
s[1:-1:2]

3    b
2    d
dtype: object

#### 【चेतावनी】सूचकांक प्रकारों पर निर्देश

यदि आप परेशानी में नहीं पड़ना चाहते हैं, तो कृपया शुद्ध फ़्लोटिंग पॉइंट और किसी भी मिश्रित प्रकार (स्ट्रिंग्स, पूर्णांक, फ़्लोटिंग पॉइंट प्रकार इत्यादि का मिश्रण) को इंडेक्स के रूप में उपयोग न करें, अन्यथा आपको त्रुटि मिल सकती है या अप्रत्याशित परिणाम मिल सकते हैं विशिष्ट ऑपरेशन में, और वास्तविक डेटा विश्लेषण में ऐसा करने के लिए कोई प्रेरणा नहीं है।

#### 【अंत】

### 3.लोक इंडेक्सर

हमने पहले 'डेटाफ़्रेम' के कॉलम चुनने के बारे में बात की थी, और अब हम टेबल के लिए इसकी पंक्ति चयन पर चर्चा करेंगेतों, दो इंडेक्सर हैं, एक **तत्व**-आधारित `loc` इंडेक्सर है, और दूसरा **स्थिति**-आधारित `iloc` इंडेक्सर है।

`loc` इंडेक्सर का सामान्य रूप `loc[*, *]` है, जहां पहला `*` पंक्तियों के चयन का प्रतिनिधित्व करता है, और दूसरा `*` कॉलम के चयन का प्रतिनिधित्व करता है यदि दूसरी स्थिति छोड़ दी जाती है `loc[*]` के रूप में लिखा गया है, यह `*` पंक्तियों के चयन को संदर्भित करता है, उनमें से `*` स्थिति में पांच प्रकार की कानूनी वस्तुएं हैं, अर्थात्: एकल तत्व, तत्व लिसटी, एलिमेंट स्लाइस, बूलियन सूची और फ़ंक्शन, जिसे नीचे बारी-बारी से समझाया जाएगा।

संबंधित ऑपरेशन को प्रदर्शित करने के लिए, हम पहले `नाम` कॉलम को एक इंडेक्स के रूप में सेट करने के लिए `set_index` विधि का उपयोग करते हैं। इस फ़ंक्शन के अन्य उपयोग मल्टी-लेवल इंडेक्सिंग पर अध्याय में पेश किए जाएंगे।

In [14]:
df_demo = df.set_index('Name')
df_demo.head()

School      Grade  Gender  Weight  \
Name                                                                       
Gaopeng Yang    Shanghai Jiao Tong University   Freshman  Female    46.0   
Changqiang You              Peking University   Freshman    Male    70.0   
Mei Sun         Shanghai Jiao Tong University     Senior    Male    89.0   
Xiaojuan Sun                 Fudan University  Sophomore  Female    41.0   
Gaojuan You                  Fudan University  Sophomore    Male    74.0   

               Transfer  
Name                     
Gaopeng Yang          N  
Changqiang You        N  
Mei Sun               N  
Xiaojuan Sun          N  
Gaojuan You           N

【a】`*` एक एकल तत्व है

इस समय, सीधे संबंधित पंक्ति या कॉलम को हटा दें यदि तत्व को सूचकांक में दोहराया जाता है, तो परिणाम `डेटाफ़्रेम` है, अन्यथा यह `श्रृंखला` है:

In [15]:
df_demo.loc['Qiang Sun'] # 多个人叫此名字

School      Grade  Gender  Weight Transfer
Name                                                                        
Qiang Sun            Tsinghua University     Junior  Female    53.0        N
Qiang Sun            Tsinghua University  Sophomore  Female    40.0        N
Qiang Sun  Shanghai Jiao Tong University     Junior  Female     NaN        N

In [16]:
df_demo.loc['Quan Zhao'] # 名字唯一

School      Shanghai Jiao Tong University
Grade                              Junior
Gender                             Female
Weight                               53.0
Transfer                                N
Name: Quan Zhao, dtype: object

पंक्तियों और स्तंभों का एक साथ चयन करना भी संभव है:

In [17]:
df_demo.loc['Qiang Sun', 'School'] # 返回Series

Name
Qiang Sun              Tsinghua University
Qiang Sun              Tsinghua University
Qiang Sun    Shanghai Jiao Tong University
Name: School, dtype: object

In [18]:
df_demo.loc['Quan Zhao', 'School'] # 返回单个元素

'Shanghai Jiao Tong University'

【बी】`*` एक तत्व सूची है

इस समय, सूची में सभी तत्व मानों से संबंधित पंक्तियों या स्तंभों को हटा दें:

In [19]:
df_demo.loc[['Qiang Sun','Quan Zhao'], ['School','Gender']]

School  Gender
Name                                            
Qiang Sun            Tsinghua University  Female
Qiang Sun            Tsinghua University  Female
Qiang Sun  Shanghai Jiao Tong University  Female
Quan Zhao  Shanghai Jiao Tong University  Female

【c】`*` एक टुकड़ा है

पिछली 'श्रृंखला' में स्ट्रिंग इंडेक्स का उपयोग करते समय, यह उल्लेख किया गया था कि यदि प्रारंभ और अंत वर्ण अद्वितीय मान हैं, तो एक स्लाइस का उपयोग किया जा सकता है और दोनों समापन बिंदु शामिल किए गए हैं, यदि वे अद्वितीय नहीं हैं, तो एक त्रुटि की सूचना दी जाती है:

In [20]:
df_demo.loc['Gaojuan You':'Gaoqiang Qian', 'School':'Gender']

School      Grade  Gender
Name                                                           
Gaojuan You                 Fudan University  Sophomore    Male
Xiaoli Qian              Tsinghua University   Freshman  Female
Qiang Chu      Shanghai Jiao Tong University   Freshman  Female
Gaoqiang Qian            Tsinghua University     Junior  Female

ध्यान दें कि यदि `डेटाफ़्रेम` पूर्णांक अनुक्रमणिका का उपयोग करता है, तो पूर्णांक स्लाइसिंग की आवश्यकताएं उपरोक्त स्ट्रिंग इंडेक्स के समान हैं, अर्थात, अंतिम बिंदुओं सहित **तत्व** स्लाइसिंग, और किसी भी डुप्लिकेट मान की अनुमति नहीं है आरंभ और अंत बिंदु.

In [21]:
df_loc_slice_demo = df_demo.copy()
df_loc_slice_demo.index = range(df_demo.shape[0],0,-1)
df_loc_slice_demo.loc[5:3]

School   Grade  Gender  Weight Transfer
5               Fudan University  Junior  Female    46.0        N
4            Tsinghua University  Senior  Female    50.0        N
3  Shanghai Jiao Tong University  Senior  Female    45.0        N

In [22]:
df_loc_slice_demo.loc[3:5] # 没有返回，说明不是整数位置切片

Empty DataFrame
Columns: [School, Grade, Gender, Weight, Transfer]
Index: []

[डी] `*` एक बूलियन सूची है

वास्तविक डेटा प्रोसेसिंग में, शर्तों के आधार पर पंक्तियों को फ़िल्टर करना बहुत आम है, यहां `loc` को पास की गई बूलियन सूची की लंबाई `DataFrame` के समान है, और उन स्थानों के अनुरूप पंक्तियों की लंबाई `True` होगी। चयनित किया जाएगा, और `गलत` हटा दिया जाएगा।

उदाहरण के लिए, 70 किलोग्राम से अधिक वजन वाले छात्रों का चयन करें:

In [23]:
df_demo.loc[df_demo.Weight>70].head()

School      Grade Gender  Weight  \
Name                                                                     
Mei Sun        Shanghai Jiao Tong University     Senior   Male    89.0   
Gaojuan You                 Fudan University  Sophomore   Male    74.0   
Xiaopeng Zhou  Shanghai Jiao Tong University   Freshman   Male    74.0   
Xiaofeng Sun             Tsinghua University     Senior   Male    71.0   
Qiang Zheng    Shanghai Jiao Tong University     Senior   Male    87.0   

              Transfer  
Name                    
Mei Sun              N  
Gaojuan You          N  
Xiaopeng Zhou        N  
Xiaofeng Sun         N  
Qiang Zheng          N

ऊपर उल्लिखित पारित तत्वों की सूची को 'isin' विधि द्वारा लौटाई गई बूलियन सूची का उपयोग करके समकक्ष रूप से भी लिखा जा सकता है, उदाहरण के लिए, नए और वरिष्ठ नागरिकों की सभी जानकारी का चयन करने के लिए:

In [24]:
df_demo.loc[df_demo.Grade.isin(['Freshman', 'Senior'])].head()

School     Grade  Gender  Weight  \
Name                                                                      
Gaopeng Yang    Shanghai Jiao Tong University  Freshman  Female    46.0   
Changqiang You              Peking University  Freshman    Male    70.0   
Mei Sun         Shanghai Jiao Tong University    Senior    Male    89.0   
Xiaoli Qian               Tsinghua University  Freshman  Female    51.0   
Qiang Chu       Shanghai Jiao Tong University  Freshman  Female    52.0   

               Transfer  
Name                     
Gaopeng Yang          N  
Changqiang You        N  
Mei Sun               N  
Xiaoli Qian           N  
Qiang Chu             N

यौगिक स्थितियों के लिए, आप इसे प्राप्त करने के लिए | (या), और (और), और ~ (नकारात्मक) के संयोजन का उपयोग कर सकते हैं, उदाहरण के लिए, फुडन विश्वविद्यालय में वरिष्ठ छात्रों का चयन करें जिनका वजन 70 किलोग्राम से अधिक है, या पुरुष छात्र हैं पेकिंग विश्वविद्यालय जिनका वजन 80 किलोग्राम से अधिक है लेकिन वे वरिष्ठ नहीं हैं:

In [25]:
condition_1_1 = df_demo.School == 'Fudan University'
condition_1_2 = df_demo.Grade == 'Senior'
condition_1_3 = df_demo.Weight > 70
condition_1 = condition_1_1 & condition_1_2 & condition_1_3
condition_2_1 = df_demo.School == 'Peking University'
condition_2_2 = df_demo.Grade == 'Senior'
condition_2_3 = df_demo.Weight > 80
condition_2 = condition_2_1 & (~condition_2_2) & condition_2_3
df_demo.loc[condition_1 | condition_2]

School     Grade Gender  Weight Transfer
Name                                                               
Qiang Han       Peking University  Freshman   Male    87.0        N
Chengpeng Zhou   Fudan University    Senior   Male    81.0        N
Changpeng Zhao  Peking University  Freshman   Male    83.0        N
Chengpeng Qian   Fudan University    Senior   Male    73.0        Y

#### 【अभ्यास】

`select_dtypes` एक व्यावहारिक फ़ंक्शन है जो किसी तालिका से संबंधित प्रकार के कॉलम का चयन कर सकता है। सभी संख्यात्मक कॉलम का चयन करने के लिए, बस `.select_dtypes('number')` का उपयोग करें। कृपया `dtypes` के साथ संयुक्त बूलियन सूची चयन विधि का उपयोग करें इस फ़ंक्शन को `learn_pandas` डेटासेट पर लागू करने के लिए `DataFrame` की विशेषता।

#### 【अंत】

【e】`*` एक फ़ंक्शन है

यहां फ़ंक्शन को ऊपर उल्लिखित चार कानूनी रूपों में से एक को वापस करना होगा, और फ़ंक्शन का इनपुट मान `DataFr' हैame` स्वयं। यह मानते हुए कि यह अभी भी ऊपर उल्लिखित यौगिक स्थिति स्क्रीनिंग का उदाहरण है, तर्क को एक फ़ंक्शन में लिखा जा सकता है और फिर वापस किया जा सकता है। यह ध्यान दिया जाना चाहिए कि फ़ंक्शन का औपचारिक पैरामीटर `x` अनिवार्य रूप से `df_demo` है।

In [26]:
def condition(x):
    condition_1_1 = x.School == 'Fudan University'
    condition_1_2 = x.Grade == 'Senior'
    condition_1_3 = x.Weight > 70
    condition_1 = condition_1_1 & condition_1_2 & condition_1_3
    condition_2_1 = x.School == 'Peking University'
    condition_2_2 = x.Grade == 'Senior'
    condition_2_3 = x.Weight > 80
    condition_2 = condition_2_1 & (~condition_2_2) & condition_2_3
    result = condition_1 | condition_2
    return result
df_demo.loc[condition]

School     Grade Gender  Weight Transfer
Name                                                               
Qiang Han       Peking University  Freshman   Male    87.0        N
Chengpeng Zhou   Fudan University    Senior   Male    81.0        N
Changpeng Zhao  Peking University  Freshman   Male    83.0        N
Chengpeng Qian   Fudan University    Senior   Male    73.0        Y

इसके अलावा, लैम्ब्डा एक्सप्रेशन भी समर्थित हैं, और उनका रिटर्न मान भी पहले बताए गए चार रूपों में से एक होना चाहिए:

In [27]:
df_demo.loc[lambda x:'Quan Zhao', lambda x:'Gender']

'Female'

चूंकि फ़ंक्शन `स्टार्ट: एंड: स्टेप` जैसे स्लाइस प्रारूप को वापस नहीं कर सकता है, इसलिए इसे वापस करते समय स्लाइस को `स्लाइस` ऑब्जेक्ट के साथ लपेटा जाना चाहिए:

In [28]:
df_demo.loc[lambda x: slice('Gaojuan You', 'Gaoqiang Qian')]

School      Grade  Gender  Weight  \
Name                                                                      
Gaojuan You                 Fudan University  Sophomore    Male    74.0   
Xiaoli Qian              Tsinghua University   Freshman  Female    51.0   
Qiang Chu      Shanghai Jiao Tong University   Freshman  Female    52.0   
Gaoqiang Qian            Tsinghua University     Junior  Female    50.0   

              Transfer  
Name                    
Gaojuan You          N  
Xiaoli Qian          N  
Qiang Chu            N  
Gaoqiang Qian        N

अंत में, यह बताया जाना चाहिए कि लोक इंडेक्स का उपयोग श्रृंखला के लिए भी किया जा सकता है। अपनाया गया सिद्धांत बिल्कुल वैसा ही है जैसा कि डेटाफ़्रेम में पंक्ति फ़िल्टरिंग के लिए उपयोग किया जाने वाला लोक [*] है, इसलिए मैं यहां विवरण में नहीं जाऊंगा।

#### 【चेतावनी】जंजीरयुक्त असाइनमेंट का उपयोग न करें

किसी तालिका या अनुक्रम में मान निर्दिष्ट करते समय, आपको इंडेक्सर्स की एक परत का उपयोग करने के बाद सीधे असाइनमेंट ऑपरेशन करना चाहिए, ऐसा इसलिए है क्योंकि एकाधिक इंडेक्स के बाद, असाइनमेंट अस्थायी रूप से लौटाई गई 'कॉपी' कॉपी और तत्व को सौंपा जाता है वास्तव में संशोधित नहीं किया गया है, इस प्रकार 'SettingWithCopyWarning' चेतावनी की रिपोर्ट की गई है, उदाहरण के लिए:

In [29]:
df_chain = pd.DataFrame([[0,0],[1,0],[-1,0]], columns=list('AB'))
df_chain
import warnings
with warnings.catch_warnings():
    warnings.filterwarnings('error')
    try:
        df_chain[df_chain.A!=0].B = 1 # 使用方括号列索引后，再使用点的列索引
    except Warning as w:
        Warning_Msg = w
print(Warning_Msg)
df_chain


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


A  B
0  0  0
1  1  0
2 -1  0

In [30]:
df_chain.loc[df_chain.A!=0,'B'] = 1
df_chain

A  B
0  0  0
1  1  1
2 -1  1

#### 【अंत】

### 4. आईलोक इंडेक्सर

`iloc` का उपयोग बिल्कुल `loc` के समान है, सिवाय इसके कि यह फ़िल्टरिंग की स्थिति पर आधारित है। संबंधित `*` स्थिति पर पांच प्रकार की कानूनी वस्तुएं भी हैं, अर्थात्: पूर्णांक, पूर्णांक सूची, पूर्णांक। स्लाइस, बूलियन सूची और फ़ंक्शन। फ़ंक्शन का रिटर्न मान उपरोक्त चार प्रकार की कानूनी वस्तुओं में से एक होना चाहिए, और इसका इनपुट भी `डेटाफ़्रेम` ही है।

In [31]:
df_demo.iloc[1, 1] # 第二行第二列

'Freshman'

In [32]:
df_demo.iloc[[0, 1], [0, 1]] # 前两行前两列

School     Grade
Name                                                   
Gaopeng Yang    Shanghai Jiao Tong University  Freshman
Changqiang You              Peking University  Freshman

In [33]:
df_demo.iloc[1: 4, 2:4] # 切片不包含结束端点

Gender  Weight
Name                          
Changqiang You    Male    70.0
Mei Sun           Male    89.0
Xiaojuan Sun    Female    41.0

In [34]:
df_demo.iloc[lambda x: slice(1, 4)] # 传入切片为返回值的函数

School      Grade  Gender  Weight  \
Name                                                                       
Changqiang You              Peking University   Freshman    Male    70.0   
Mei Sun         Shanghai Jiao Tong University     Senior    Male    89.0   
Xiaojuan Sun                 Fudan University  Sophomore  Female    41.0   

               Transfer  
Name                     
Changqiang You        N  
Mei Sun               N  
Xiaojuan Sun          N

बूलियन सूचियों का उपयोग करते समय, आपको 'सीरीज़' नहीं बल्कि अनुक्रम के 'मानों' को पास करने पर विशेष ध्यान देना चाहिए, अन्यथा एक त्रुटि की सूचना दी जाएगी, इसलिए, बूलियन फ़िल्टरिंग का उपयोग करते समय, 'लोक' विधि को प्राथमिकता दी जानी चाहिए।

उदाहरण के लिए, 80 किलोग्राम से अधिक वजन वाले छात्रों का चयन करने के लिए:

In [35]:
df_demo.iloc[(df_demo.Weight>80).values].head()

School      Grade Gender  Weight  \
Name                                                                      
Mei Sun         Shanghai Jiao Tong University     Senior   Male    89.0   
Qiang Zheng     Shanghai Jiao Tong University     Senior   Male    87.0   
Qiang Han                   Peking University   Freshman   Male    87.0   
Chengpeng Zhou               Fudan University     Senior   Male    81.0   
Feng Han        Shanghai Jiao Tong University  Sophomore   Male    82.0   

               Transfer  
Name                     
Mei Sun               N  
Qiang Zheng           N  
Qiang Han             N  
Chengpeng Zhou        N  
Feng Han              N

`सीरीज़` के लिए, आप संबंधित स्थिति में मूल्य या परिणाम वापस करने के लिए `iloc` का भी उपयोग कर सकते हैं:

In [36]:
df_demo.School.iloc[1]

'Peking University'

In [37]:
df_demo.School.iloc[1:5:2]

Name
Changqiang You    Peking University
Xiaojuan Sun       Fudan University
Name: School, dtype: object

### 5. क्वेरी विधि

`पांडा` में, आप डेटा को क्वेरी करने के लिए `क्वेरी` विधि में एक स्ट्रिंग-जैसी क्वेरी अभिव्यक्ति पास कर सकते हैं। जटिल अनुक्रमण करते समय अभिव्यक्ति के निष्पादन परिणाम को एक बूलियन सूची वापस करनी होगी, इस पुनर्प्राप्ति विधि को बार-बार उपयोग करने की आवश्यकता नहीं है `डेटाफ़्रेम` नाम सामान्य विधि की तरह कॉलम नाम को संदर्भित करता है, जो आम तौर पर पठनीयता को कम किए बिना कोड की लंबाई को कम करता है।

उदाहरण के लिए, `loc` अनुभाग में यौगिक स्थिति क्वेरी उदाहरण को फिर से लिखा जा सकता हैn इस प्रकार है:

In [38]:
df.query('((School == "Fudan University")&'
         ' (Grade == "Senior")&'
         ' (Weight > 70))|'
         '((School == "Peking University")&'
         ' (Grade != "Senior")&'
         ' (Weight > 80))')

School     Grade            Name Gender  Weight Transfer
38   Peking University  Freshman       Qiang Han   Male    87.0        N
66    Fudan University    Senior  Chengpeng Zhou   Male    81.0        N
99   Peking University  Freshman  Changpeng Zhao   Male    83.0        N
131   Fudan University    Senior  Chengpeng Qian   Male    73.0        Y

क्वेरी अभिव्यक्ति में, डेटाफ़्रेम से सभी कॉलम नाम उपयोगकर्ता के लिए पंजीकृत हैं। श्रृंखला से संबंधित सभी तरीकों को कॉल किया जा सकता है, जो सामान्य फ़ंक्शन कॉल से अलग नहीं है, उदाहरण के लिए, उन छात्रों से पूछताछ करना जिनका वजन माध्य से अधिक है।

In [39]:
df.query('Weight > Weight.mean()').head()

School      Grade            Name  Gender  Weight  \
1               Peking University   Freshman  Changqiang You    Male    70.0   
2   Shanghai Jiao Tong University     Senior         Mei Sun    Male    89.0   
4                Fudan University  Sophomore     Gaojuan You    Male    74.0   
10  Shanghai Jiao Tong University   Freshman   Xiaopeng Zhou    Male    74.0   
14            Tsinghua University     Senior    Xiaomei Zhou  Female    57.0   

   Transfer  
1         N  
2         N  
4         N  
10        N  
14        N

#### 【नोट】क्वेरी में रिक्त स्थान के साथ कॉलम नाम उद्धृत करें

रिक्त स्थान वाले कॉलम नामों के लिए, आपको उन्हें उद्धृत करने के लिए `` `col name` `` का उपयोग करना होगा।

#### 【अंत】

साथ ही, पठनीयता में सुधार करने में सहायता के लिए `क्वेरी` में कई अंग्रेजी शाब्दिक उपयोग पंजीकृत हैं, जैसे: `या, और, या, इन, नॉट इन` उदाहरण के लिए, उन पुरुष छात्रों को फ़िल्टर करें जो नए या दूसरे वर्ष के छात्र नहीं हैं:

In [40]:
df.query('(Grade not in ["Freshman", "Sophomore"]) and (Gender == "Male")').head()

School   Grade           Name Gender  Weight  \
2   Shanghai Jiao Tong University  Senior        Mei Sun   Male    89.0   
16            Tsinghua University  Junior  Xiaoqiang Qin   Male    68.0   
17            Tsinghua University  Junior      Peng Wang   Male    65.0   
18            Tsinghua University  Senior   Xiaofeng Sun   Male    71.0   
21  Shanghai Jiao Tong University  Senior  Xiaopeng Shen   Male    62.0   

   Transfer  
2         N  
16        N  
17        N  
18        N  
21      NaN

इसके अलावा, जब किसी सूची के साथ तुलना एक स्ट्रिंग में दिखाई देती है, तो क्रमशः `==` और `!=` इंगित करता है कि तत्व सूची में दिखाई देता है और सूची में प्रकट नहीं होता है, जो `in` और `not के बराबर है उदाहरण के लिए, सभी जूनियर और सीनियर छात्रों से पूछताछ करने के लिए:

In [41]:
df.query('Grade == ["Junior", "Senior"]').head()

School   Grade           Name  Gender  Weight  \
2   Shanghai Jiao Tong University  Senior        Mei Sun    Male    89.0   
7             Tsinghua University  Junior  Gaoqiang Qian  Female    50.0   
9               Peking University  Junior        Juan Xu  Female     NaN   
11            Tsinghua University  Junior    Xiaoquan Lv  Female    43.0   
12  Shanghai Jiao Tong University  Senior       Peng You  Female    48.0   

   Transfer  
2         N  
7         N  
9         N  
11        N  
12      NaN

`क्वेरी` में स्ट्रिंग के लिए, यदि आप किसी बाहरी चर को संदर्भित करना चाहते हैं, तो बस चर नाम से पहले `@` प्रतीक जोड़ें, उदाहरण के लिए, उन छात्रों को पुनः प्राप्त करने के लिए जिनका वजन 70 किलोग्राम और 80 किलोग्राम के बीच है:

In [42]:
low, high =70, 80
df.query('(Weight >= @low) & (Weight <= @high)').head()

School      Grade            Name Gender  Weight  \
1               Peking University   Freshman  Changqiang You   Male    70.0   
4                Fudan University  Sophomore     Gaojuan You   Male    74.0   
10  Shanghai Jiao Tong University   Freshman   Xiaopeng Zhou   Male    74.0   
18            Tsinghua University     Senior    Xiaofeng Sun   Male    71.0   
35              Peking University   Freshman      Gaoli Zhao   Male    78.0   

   Transfer  
1         N  
4         N  
10        N  
18        N  
35        N

### 6. यादृच्छिक नमूनाकरण

यदि आप 'डेटाफ़्रेम' की प्रत्येक पंक्ति को एक नमूना मानते हैं, या प्रत्येक कॉलम को एक सुविधा के रूप में मानते हैं, और संपूर्ण 'डेटाफ़्रेम' को एक जनसंख्या के रूप में मानते हैं, तो आप कभी-कभी बड़े पैमाने पर नमूने या सुविधाओं का नमूना लेने के लिए 'नमूना' फ़ंक्शन का उपयोग कर सकते हैं डेटा सेट, आप डेटा के अनुमानित वितरण को समझने के लिए सांख्यिकीय विशेषताओं की गणना करना चाहते हैं, लेकिन इसमें बहुत समय लगता है, क्योंकि कई सांख्यिकीय विशेषताएं सांख्यिकीय विशेषताओं के निष्पक्ष अनुमान हैंसमान संभावना और बिना किसी प्रतिस्थापन के सरल यादृच्छिक नमूने की स्थिति के तहत जनसंख्या का, जैसे नमूना माध्य और जनसंख्या माध्य, आप अनुमानित अनुमान लगाने के लिए पहले पूरी तालिका से एक हिस्सा निकाल सकते हैं।

`नमूना` फ़ंक्शन में मुख्य पैरामीटर `एन, अक्ष, फ़्रेक, प्रतिस्थापित, वजन` हैं। पहले तीन नमूनों की संख्या, नमूने की दिशा (पंक्तियों के लिए 0 और स्तंभों के लिए 1), और नमूना अनुपात को संदर्भित करते हैं। (0.3 का मतलब है कि 30% नमूने लिए गए हैंआबादी)।

`रिप्लेस` और `वेट` क्रमशः प्रत्येक नमूने को प्रतिस्थापित करने और नमूना लेने की सापेक्ष संभावना को संदर्भित करते हैं। जब `रिप्लेस = ट्रू` होता है, तो इसका मतलब प्रतिस्थापन के साथ नमूनाकरण होता है, उदाहरण के लिए, निम्नलिखित निर्मित `df_sample` के सापेक्ष आकार का उपयोग करता है प्रतिस्थापन के साथ नमूनाकरण करने की नमूना संभावना के रूप में `मूल्य`, और नमूनों की संख्या 3 है।

In [43]:
df_sample = pd.DataFrame({'id': list('abcde'), 'value': [1, 2, 3, 4, 90]})
df_sample

id  value
0  a      1
1  b      2
2  c      3
3  d      4
4  e     90

In [44]:
df_sample.sample(3, replace = True, weights = df_sample.value)

id  value
4  e     90
4  e     90
4  e     90

## 2. बहुस्तरीय सूचकांक
### 1. बहु-स्तरीय सूचकांक और इसकी तालिका संरचना

बहु-स्तरीय सूचकांक के साथ 'डेटाफ़्रेम' संरचना को अधिक स्पष्ट रूप से चित्रित करने के लिए, नीचे एक नई तालिका बनाई गई है। पाठक यहां निर्माण विधि को अनदेखा कर सकते हैं, जिसे अनुभाग 4 में अधिक विस्तार से बताया जाएगा।

In [45]:
np.random.seed(0)
multi_index = pd.MultiIndex.from_product([list('ABCD'), df.Gender.unique()], names=('School', 'Gender'))
multi_column = pd.MultiIndex.from_product([['Height', 'Weight'], df.Grade.unique()], names=('Indicator', 'Grade'))
df_multi = pd.DataFrame(np.c_[(np.random.randn(8,4)*5 + 163).tolist(), (np.random.randn(8,4)*5 + 65).tolist()],
                        index = multi_index, columns = multi_column).round(1)
df_multi

Indicator       Height                           Weight                   \
Grade         Freshman Senior Sophomore Junior Freshman Senior Sophomore   
School Gender                                                              
A      Female    171.8  165.0     167.9  174.2     60.6   55.1      63.3   
       Male      172.3  158.1     167.8  162.2     71.2   71.0      63.1   
B      Female    162.5  165.1     163.7  170.3     59.8   57.9      56.5   
       Male      166.8  163.6     165.2  164.7     62.5   62.8      58.7   
C      Female    170.5  162.0     164.6  158.7     56.9   63.9      60.5   
       Male      150.2  166.3     167.3  159.3     62.4   59.1      64.9   
D      Female    174.3  155.7     163.2  162.1     65.3   66.5      61.8   
       Male      170.7  170.3     163.8  164.9     61.6   63.2      60.9   

Indicator             
Grade         Junior  
School Gender         
A      Female   65.8  
       Male     63.5  
B      Female   74.8  
       Male     68.9  
C      Female   66.9  
       Male     67.1  
D      Female   63.2  
       Male     56.4

नीचे दिया गया चित्र `डेटाफ़्रेम` की संरचना को चिह्नित करने के लिए रंगों का उपयोग करता है। एकल-परत सूचकांक तालिका की तरह, इसके तीन भाग हैं: तत्व मान, पंक्ति सूचकांक और स्तंभ सूचकांक। इनमें से, पंक्ति सूचकांक और स्तंभ सूचकांक दोनों हैं मल्टीइंडेक्स` प्रकार, सिवाय इसके कि ** इंडेक्स में एक तत्व सिंगल-लेयर इंडेक्स में स्केलर के बजाय एक टुपल ** है, उदाहरण के लिए, पंक्ति इंडेक्स का चौथा तत्व `("बी", "पुरुष") है। `, और कॉलम इंडेक्स का दूसरा तत्व `('ऊंचाई', 'वरिष्ठ') है, इसे यहां नोट किया जाना चाहिएजब एक ही मान बाहरी परत में लगातार दिखाई देता है, तो पहले के बाद पहला छिपाया जाएगा और प्रदर्शित किया जाएगा, जिससे परिणाम अधिक पठनीय हो जाएगा।

<img src='../source/_static/multi_index.png' चौड़ाई='50%'>

सिंगल-लेयर इंडेक्स के समान, `मल्टीइंडेक्स` में भी एक नाम विशेषता है, `स्कूल` और `लिंग` क्रमशः तालिका की पहली और दूसरी परत पंक्ति इंडेक्स के नाम और `संकेतक` से मेल खाते हैं। ` और `ग्रेड` पहले और दूसरे के नाम से मेल खाते हैंक्रमशः ऑन-लेयर कॉलम इंडेक्स।

सूचकांक विशेषताओं के नाम और मान क्रमशः `नाम` और `मान` के माध्यम से उपलब्ध हैं:

In [46]:
df_multi.index.names

FrozenList(['School', 'Gender'])

In [47]:
df_multi.columns.names

FrozenList(['Indicator', 'Grade'])

In [48]:
df_multi.index.values

array([('A', 'Female'), ('A', 'Male'), ('B', 'Female'), ('B', 'Male'),
       ('C', 'Female'), ('C', 'Male'), ('D', 'Female'), ('D', 'Male')],
      dtype=object)

In [49]:
df_multi.columns.values

array([('Height', 'Freshman'), ('Height', 'Senior'),
       ('Height', 'Sophomore'), ('Height', 'Junior'),
       ('Weight', 'Freshman'), ('Weight', 'Senior'),
       ('Weight', 'Sophomore'), ('Weight', 'Junior')], dtype=object)

यदि आप एक निश्चित स्तर का सूचकांक प्राप्त करना चाहते हैं, तो आपको इसे `get_level_values` के माध्यम से प्राप्त करना होगा:

In [50]:
df_multi.index.get_level_values(0)

Index(['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D'], dtype='object', name='School')

हालाँकि, इंडेक्स के लिए, चाहे एकल-स्तरीय या बहु-स्तरीय, उपयोगकर्ता तत्वों को `index_obj[0] = item` द्वारा संशोधित नहीं कर सकते हैं, न ही वे `index_name[0] = new_name` द्वारा नामों को संशोधित कर सकते हैं। इन्हें कैसे संशोधित किया जाए, इसका विषय विशेषताओं पर धारा 3 में चर्चा की जाएगी।

### 2. मल्टी-लेवल इंडेक्स में लोक इंडेक्सर

संरचना से परिचित होने के बाद, अब मूल तालिका पर लौटें और स्कूल और ग्रेड को अनुक्रमित के रूप में सेट करें। इस समय, पंक्तियाँ बहु-स्तरीय अनुक्रमित हैं और कॉलम एकल-स्तरीय अनुक्रमित हैंडिफ़ॉल्ट कॉलम इंडेक्स में कोई नाम नहीं है, अभी के आंकड़े में `इंडिकेटर` और `ग्रेड` से संबंधित इंडेक्स नाम की स्थिति खाली है।

In [51]:
df_multi = df.set_index(['School', 'Grade'])
df_multi.head()

Name  Gender  Weight  \
School                        Grade                                       
Shanghai Jiao Tong University Freshman     Gaopeng Yang  Female    46.0   
Peking University             Freshman   Changqiang You    Male    70.0   
Shanghai Jiao Tong University Senior            Mei Sun    Male    89.0   
Fudan University              Sophomore    Xiaojuan Sun  Female    41.0   
                              Sophomore     Gaojuan You    Male    74.0   

                                        Transfer  
School                        Grade               
Shanghai Jiao Tong University Freshman         N  
Peking University             Freshman         N  
Shanghai Jiao Tong University Senior           N  
Fudan University              Sophomore        N  
                              Sophomore        N

चूंकि मल्टी-लेवल इंडेक्स में एकल तत्व एक टपल है, पहले खंड में पेश किए गए ``loc`` और ``iloc`` तरीकों को पूरी तरह से कॉपी किया जा सकता है, बस स्केलर स्थिति को संबंधित टपल के साथ बदलें।

टुपल्स या एकल टुपल की सूची में गुजरते समय, या एक फ़ंक्शन जो पहले या बाद वाले को लौटाता है, आपको प्रदर्शन चेतावनियों से बचने के लिए पहले इंडेक्स को सॉर्ट करना होगा:

In [52]:
with warnings.catch_warnings():
    warnings.filterwarnings('error')
    try:
        df_multi.loc[('Fudan University', 'Junior')].head()
    except Warning as w:
        Warning_Msg = w
Warning_Msg

pandas.errors.PerformanceWarning('indexing past lexsort depth may impact performance.')

In [53]:
df_sorted = df_multi.sort_index()
df_sorted.loc[('Fudan University', 'Junior')].head()

Name  Gender  Weight Transfer
School           Grade                                         
Fudan University Junior      Yanli You  Female    48.0        N
                 Junior  Chunqiang Chu    Male    72.0        N
                 Junior   Changfeng Lv    Male    76.0        N
                 Junior     Yanjuan Lv  Female    49.0      NaN
                 Junior  Gaoqiang Zhou  Female    43.0        N

In [54]:
df_sorted.loc[[('Fudan University', 'Senior'), ('Shanghai Jiao Tong University', 'Freshman')]].head()

Name  Gender  Weight Transfer
School           Grade                                           
Fudan University Senior  Chengpeng Zheng  Female    38.0        N
                 Senior        Feng Zhou  Female    47.0        N
                 Senior        Gaomei Lv  Female    34.0        N
                 Senior        Chunli Lv  Female    56.0        N
                 Senior   Chengpeng Zhou    Male    81.0        N

In [55]:
df_sorted.loc[df_sorted.Weight > 70].head() # 布尔列表也是可用的

Name Gender  Weight Transfer
School           Grade                                           
Fudan University Freshman       Feng Wang   Male    74.0        N
                 Junior     Chunqiang Chu   Male    72.0        N
                 Junior      Changfeng Lv   Male    76.0        N
                 Senior    Chengpeng Zhou   Male    81.0        N
                 Senior    Chengpeng Qian   Male    73.0        Y

In [56]:
df_sorted.loc[lambda x:('Fudan University','Junior')].head()

Name  Gender  Weight Transfer
School           Grade                                         
Fudan University Junior      Yanli You  Female    48.0        N
                 Junior  Chunqiang Chu    Male    72.0        N
                 Junior   Changfeng Lv    Male    76.0        N
                 Junior     Yanjuan Lv  Female    49.0      NaN
                 Junior  Gaoqiang Zhou  Female    43.0        N

स्लाइस का उपयोग करते समय, कृपया ध्यान दें कि एकल-स्तरीय इंडेक्स में, जब तक स्लाइस एंडपॉइंट तत्व अद्वितीय होते हैं, तब तक स्लाइसिंग की जा सकती है, हालांकि, मल्टी-लेवल इंडेक्स में, भले ही टुपल्स इंडेक्स में बार-बार दिखाई देते हों। स्लाइसिंग का उपयोग करने से पहले उन्हें क्रमबद्ध किया जाना चाहिए, अन्यथा एक त्रुटि की सूचना दी जाएगी:

In [57]:
try:
    df_multi.loc[('Fudan University', 'Senior'):].head()
except Exception as e:
    Err_Msg = e
Err_Msg

pandas.errors.UnsortedIndexError('Key length (2) was greater than MultiIndex lexsort depth (0)')

In [58]:
df_sorted.loc[('Fudan University', 'Senior'):].head()

Name  Gender  Weight Transfer
School           Grade                                           
Fudan University Senior  Chengpeng Zheng  Female    38.0        N
                 Senior        Feng Zhou  Female    47.0        N
                 Senior        Gaomei Lv  Female    34.0        N
                 Senior        Chunli Lv  Female    56.0        N
                 Senior   Chengpeng Zhou    Male    81.0        N

In [59]:
df_unique = df.drop_duplicates(subset=['School','Grade']).set_index(['School', 'Grade'])
df_unique.head()

Name  Gender  Weight  \
School                        Grade                                       
Shanghai Jiao Tong University Freshman     Gaopeng Yang  Female    46.0   
Peking University             Freshman   Changqiang You    Male    70.0   
Shanghai Jiao Tong University Senior            Mei Sun    Male    89.0   
Fudan University              Sophomore    Xiaojuan Sun  Female    41.0   
Tsinghua University           Freshman      Xiaoli Qian  Female    51.0   

                                        Transfer  
School                        Grade               
Shanghai Jiao Tong University Freshman         N  
Peking University             Freshman         N  
Shanghai Jiao Tong University Senior           N  
Fudan University              Sophomore        N  
Tsinghua University           Freshman         N

In [60]:
try:
    df_unique.loc[('Fudan University', 'Senior'):].head()
except Exception as e:
    Err_Msg = e
Err_Msg

pandas.errors.UnsortedIndexError('Key length (2) was greater than MultiIndex lexsort depth (0)')

In [61]:
df_unique.sort_index().loc[('Fudan University', 'Senior'):].head()

Name  Gender  Weight Transfer
School            Grade                                              
Fudan University  Senior     Chengpeng Zheng  Female    38.0        N
                  Sophomore     Xiaojuan Sun  Female    41.0        N
Peking University Freshman    Changqiang You    Male    70.0        N
                  Junior             Juan Xu  Female     NaN        N
                  Senior          Changli Lv  Female    41.0        N

इसके अलावा, मल्टी-लेवल इंडेक्स में टुपल्स का विशेष उपयोग होता है, जिसका उपयोग तत्वों की कई परतों को क्रॉस-संयोजन और इंडेक्स करने के लिए किया जा सकता है, लेकिन साथ ही, `loc` के कॉलम को निर्दिष्ट करने की आवश्यकता होती है, और ` :` का उपयोग सभी को चुनने के लिए किया जाता है। प्रत्येक स्तर पर चुने जाने वाले तत्वों को एक सूची में संग्रहीत किया जाता है, और `loc` को पास करने का रूप `[(level_0_list,level_1_list), cols]` है, उदाहरण के लिए, यदि आप चाहें पेकिंग विश्वविद्यालय और फ़ुडन विश्वविद्यालय के सभी द्वितीय वर्ष के छात्रों और कनिष्ठों को प्राप्त करने के लिए, आप लिख सकते हैंइसे इस प्रकार से करें:

In [62]:
res = df_multi.loc[(['Peking University', 'Fudan University'], ['Sophomore', 'Junior']), :]
res.head()

Name  Gender  Weight Transfer
School            Grade                                           
Peking University Sophomore   Changmei Xu  Female    43.0        N
                  Sophomore  Xiaopeng Qin    Male     NaN        N
                  Sophomore        Mei Xu  Female    39.0        N
                  Sophomore   Xiaoli Zhou  Female    55.0        N
                  Sophomore      Peng Han  Female    34.0      NaN

In [63]:
res.shape

(33, 4)

निम्नलिखित कथन उपरोक्त के समान है, लेकिन यह अभी भी तत्वों की सूची में गुजरता है (इस मामले में टुपल्स) उनके अर्थ अलग हैं, यह दर्शाता है कि पेकिंग विश्वविद्यालय के तीसरे वर्ष के छात्रों और फुडन विश्वविद्यालय के द्वितीय वर्ष के छात्रों का चयन किया गया है:

In [64]:
res = df_multi.loc[[('Peking University', 'Junior'), ('Fudan University', 'Sophomore')]]
res.head()

Name  Gender  Weight Transfer
School            Grade                                         
Peking University Junior        Juan Xu  Female     NaN        N
                  Junior  Changjuan You  Female    47.0        N
                  Junior       Gaoli Xu  Female    48.0        N
                  Junior   Gaoquan Zhou    Male    70.0        N
                  Junior      Qiang You  Female    56.0        N

In [65]:
res.shape

(16, 4)

### 3. इंडेक्सस्लाइस ऑब्जेक्ट

ऊपर प्रस्तुत विधि केवल टुपल को संपूर्ण रूप से स्लाइस कर सकती है, लेकिन प्रत्येक परत को नहीं, भले ही इंडेक्स दोहराया न जाए। इसे बूलियन सूचियों के साथ स्लाइस को मिलाने की भी अनुमति नहीं है। `इंडेक्सस्लाइस` ऑब्जेक्ट की शुरूआत इस समस्या को हल कर सकती है `स्लाइस` ऑब्जेक्ट के दो रूप हैं, पहला `loc[idx[*,*]]` प्रकार है, और दूसरा `loc[idx[*,*],idx[*,*]]` है। प्रकार, जिसे नीचे प्रस्तुत किया जाएगा प्रदर्शन की सुविधा के लिए, निम्नलिखित एक **दा का निर्माण करता हैtaFrame` गैर-डुप्लिकेट अनुक्रमणिका के साथ:

In [66]:
np.random.seed(0)
L1,L2 = ['A','B','C'],['a','b','c']
mul_index1 = pd.MultiIndex.from_product([L1,L2],names=('Upper', 'Lower'))
L3,L4 = ['D','E','F'],['d','e','f']
mul_index2 = pd.MultiIndex.from_product([L3,L4],names=('Big', 'Small'))
df_ex = pd.DataFrame(np.random.randint(-9,10,(9,9)), index=mul_index1, columns=mul_index2)
df_ex

Big          D        E        F      
Small        d  e  f  d  e  f  d  e  f
Upper Lower                           
A     a      3  6 -9 -6 -6 -2  0  9 -5
      b     -3  3 -8 -3 -2  5  8 -4  4
      c     -1  0  7 -4  6  6 -9  9 -6
B     a      8  5 -2 -9 -8  0 -9  1 -6
      b      2  9 -7 -9 -9 -5 -4 -3 -1
      c      8  6 -5  0  1 -8 -8 -2  0
C     a     -6 -3  2  5  9 -9  5 -6  3
      b      1  2 -5 -3 -5  6 -6  3 -5
      c     -1  5  6 -6  6  4  7  8 -4

साइल ऑब्जेक्ट का उपयोग करने के लिए, आपको पहले इसे परिभाषित करना होगा:

In [67]:
idx = pd.IndexSlice

[ए] `loc[idx[*,*]]` प्रकार

इस मामले में, स्लाइसिंग की कई परतें अलग-अलग नहीं की जा सकतीं। पहला `*` पंक्तियों के चयन को इंगित करता है, और दूसरा `*` स्तंभों के चयन को इंगित करता है, जो सरल `loc` के समान है:

In [68]:
df_ex.loc[idx['C':, ('D', 'f'):]]

Big          D  E        F      
Small        f  d  e  f  d  e  f
Upper Lower                     
C     a      2  5  9 -9  5 -6  3
      b     -5 -3 -5  6 -6  3 -5
      c      6 -6  6  4  7  8 -4

इसके अतिरिक्त, बूलियन अनुक्रमों का अनुक्रमण समर्थित है:

In [69]:
df_ex.loc[idx[:'A', lambda x:x.sum()>0]] # 列和大于0

Big          D     F
Small        d  e  e
Upper Lower         
A     a      3  6  9
      b     -3  3 -4
      c     -1  0  9

【b】`loc[idx[*,*],idx[*,*]]` प्रकार

इस मामले में, स्लाइसिंग परतों में की जा सकती है, जिसमें पहला `idx` पंक्ति सूचकांक को संदर्भित करता है और दूसरा स्तंभ सूचकांक को संदर्भित करता है।

In [70]:
df_ex.loc[idx[:'A', 'b':], idx['E':, 'e':]]

Big          E     F   
Small        e  f  e  f
Upper Lower            
A     b     -2  5 -4  4
      c      6  6  9 -6

लेकिन कृपया ध्यान दें कि फ़ंक्शंस का उपयोग इस समय समर्थित नहीं है:

In [71]:
try:
    df_ex.loc[idx[:'A', lambda x: 'b'], idx['E':, 'e':]]
except Exception as e:
    Err_Msg = e
Err_Msg

KeyError(<function __main__.<lambda>(x)>)

### 4. बहुस्तरीय सूचकांक का निर्माण

मल्टी-लेवल इंडेक्स टेबल की संरचना और स्लाइसिंग का उल्लेख ऊपर किया गया था। तो, `set_index` का उपयोग करने के अलावा, अपने आप से मल्टी-लेवल इंडेक्स का निर्माण कैसे करें? आमतौर पर तीन तरीकों का उपयोग किया जाता है: `from_tuples`, from_arrays, और from_product` , जो `pd.MultiIndex` ऑब्जेक्ट के अंतर्गत सभी फ़ंक्शन हैं।

`from_tuples` का अर्थ है पास किए गए टुपल्स की सूची के अनुसार निर्माण करना:

In [72]:
my_tuple = [('a','cat'),('a','dog'),('b','cat'),('b','dog')]
pd.MultiIndex.from_tuples(my_tuple, names=['First','Second'])

MultiIndex([('a', 'cat'),
            ('a', 'dog'),
            ('b', 'cat'),
            ('b', 'dog')],
           names=['First', 'Second'])

`from_arrays` का अर्थ है पारित सूची के अनुसार संबंधित परतों की सूची बनाना:

In [73]:
my_array = [list('aabb'), ['cat', 'dog']*2]
pd.MultiIndex.from_arrays(my_array, names=['First','Second'])

MultiIndex([('a', 'cat'),
            ('a', 'dog'),
            ('b', 'cat'),
            ('b', 'dog')],
           names=['First', 'Second'])

`from_product` दी गई सूचियों के कार्टेशियन उत्पाद से एक सूची बनाता है:

In [74]:
my_list1 = ['a','b']
my_list2 = ['cat','dog']
pd.MultiIndex.from_product([my_list1, my_list2], names=['First','Second'])

MultiIndex([('a', 'cat'),
            ('a', 'dog'),
            ('b', 'cat'),
            ('b', 'dog')],
           names=['First', 'Second'])

## 3. अनुक्रमण के सामान्य तरीके
### 1. सूचकांक परतों का आदान-प्रदान और विलोपन
विनिमय प्रक्रिया को समझने में सुविधा के लिए, यहां तीन-स्तरीय सूचकांक के निर्माण का एक उदाहरण दिया गया है:

In [75]:
np.random.seed(0)
L1,L2,L3 = ['A','B'],['a','b'],['alpha','beta']
mul_index1 = pd.MultiIndex.from_product([L1,L2,L3], names=('Upper', 'Lower','Extra'))
L4,L5,L6 = ['C','D'],['c','d'],['cat','dog']
mul_index2 = pd.MultiIndex.from_product([L4,L5,L6], names=('Big', 'Small', 'Other'))
df_ex = pd.DataFrame(np.random.randint(-9,10,(8,8)), index=mul_index1,  columns=mul_index2)
df_ex

Big                 C               D            
Small               c       d       c       d    
Other             cat dog cat dog cat dog cat dog
Upper Lower Extra                                
A     a     alpha   3   6  -9  -6  -6  -2   0   9
            beta   -5  -3   3  -8  -3  -2   5   8
      b     alpha  -4   4  -1   0   7  -4   6   6
            beta   -9   9  -6   8   5  -2  -9  -8
B     a     alpha   0  -9   1  -6   2   9  -7  -9
            beta   -9  -5  -4  -3  -1   8   6  -5
      b     alpha   0   1  -8  -8  -2   0  -6  -3
            beta    2   5   9  -9   5  -6   3   1

सूचकांक स्तरों का आदान-प्रदान `स्वैपलेवल` और `रीऑर्डर_लेवल्स` द्वारा किया जाता है, पहला केवल दो स्तरों का आदान-प्रदान कर सकता है, जबकि बाद वाला किसी भी स्तर का आदान-प्रदान कर सकता है, यानी पंक्ति सूचकांक या स्तंभ सूचकांक।

In [76]:
df_ex.swaplevel(0,2,axis=1).head() # 列索引的第一层和第三层交换

Other             cat dog cat dog cat dog cat dog
Small               c   c   d   d   c   c   d   d
Big                 C   C   C   C   D   D   D   D
Upper Lower Extra                                
A     a     alpha   3   6  -9  -6  -6  -2   0   9
            beta   -5  -3   3  -8  -3  -2   5   8
      b     alpha  -4   4  -1   0   7  -4   6   6
            beta   -9   9  -6   8   5  -2  -9  -8
B     a     alpha   0  -9   1  -6   2   9  -7  -9

In [77]:
df_ex.reorder_levels([2,0,1],axis=0).head() # 列表数字指代原来索引中的层

Big                 C               D            
Small               c       d       c       d    
Other             cat dog cat dog cat dog cat dog
Extra Upper Lower                                
alpha A     a       3   6  -9  -6  -6  -2   0   9
beta  A     a      -5  -3   3  -8  -3  -2   5   8
alpha A     b      -4   4  -1   0   7  -4   6   6
beta  A     b      -9   9  -6   8   5  -2  -9  -8
alpha B     a       0  -9   1  -6   2   9  -7  -9

#### 【नोट】अक्षों के बीच सूचकांक विनिमय
इसमें केवल पंक्ति या स्तंभ सूचकांकों का आदान-प्रदान शामिल है, विभिन्न दिशाओं में सूचकांकों के आदान-प्रदान पर अध्याय 5 में चर्चा की जाएगी।
#### 【अंत】
यदि आप एक निश्चित स्तर के सूचकांक को हटाना चाहते हैं, तो आप `ड्रॉपलेवल` विधि का उपयोग कर सकते हैं:

In [78]:
df_ex.droplevel(1,axis=1)

Big                 C               D            
Other             cat dog cat dog cat dog cat dog
Upper Lower Extra                                
A     a     alpha   3   6  -9  -6  -6  -2   0   9
            beta   -5  -3   3  -8  -3  -2   5   8
      b     alpha  -4   4  -1   0   7  -4   6   6
            beta   -9   9  -6   8   5  -2  -9  -8
B     a     alpha   0  -9   1  -6   2   9  -7  -9
            beta   -9  -5  -4  -3  -1   8   6  -5
      b     alpha   0   1  -8  -8  -2   0  -6  -3
            beta    2   5   9  -9   5  -6   3   1

In [79]:
df_ex.droplevel([0,1],axis=0)

Big     C               D            
Small   c       d       c       d    
Other cat dog cat dog cat dog cat dog
Extra                                
alpha   3   6  -9  -6  -6  -2   0   9
beta   -5  -3   3  -8  -3  -2   5   8
alpha  -4   4  -1   0   7  -4   6   6
beta   -9   9  -6   8   5  -2  -9  -8
alpha   0  -9   1  -6   2   9  -7  -9
beta   -9  -5  -4  -3  -1   8   6  -5
alpha   0   1  -8  -8  -2   0  -6  -3
beta    2   5   9  -9   5  -6   3   1

### 2. सूचकांक विशेषताओं को संशोधित करें
इंडेक्स लेयर का नाम `rename_axis` के माध्यम से संशोधित किया जा सकता है। सामान्य संशोधन विधि डिक्शनरी मैपिंग में पास करना है:

In [80]:
df_ex.rename_axis(index={'Upper':'Changed_row'}, columns={'Other':'Changed_Col'}).head()

Big                       C               D            
Small                     c       d       c       d    
Changed_Col             cat dog cat dog cat dog cat dog
Changed_row Lower Extra                                
A           a     alpha   3   6  -9  -6  -6  -2   0   9
                  beta   -5  -3   3  -8  -3  -2   5   8
            b     alpha  -4   4  -1   0   7  -4   6   6
                  beta   -9   9  -6   8   5  -2  -9  -8
B           a     alpha   0  -9   1  -6   2   9  -7  -9

सूचकांक मान को `नाम बदलें` के माध्यम से संशोधित किया जा सकता है। यदि यह एक बहु-स्तरीय सूचकांक है, तो संशोधित स्तर संख्या `स्तर` को निर्दिष्ट करने की आवश्यकता है:

In [81]:
df_ex.rename(columns={'cat':'not_cat'}, level=2).head()

Big                     C                       D                
Small                   c           d           c           d    
Other             not_cat dog not_cat dog not_cat dog not_cat dog
Upper Lower Extra                                                
A     a     alpha       3   6      -9  -6      -6  -2       0   9
            beta       -5  -3       3  -8      -3  -2       5   8
      b     alpha      -4   4      -1   0       7  -4       6   6
            beta       -9   9      -6   8       5  -2      -9  -8
B     a     alpha       0  -9       1  -6       2   9      -7  -9

पारित पैरामीटर एक फ़ंक्शन भी हो सकता है, जिसका इनपुट मान सूचकांक तत्व है:

In [82]:
df_ex.rename(index=lambda x:str.upper(x), level=2).head()

Big                 C               D            
Small               c       d       c       d    
Other             cat dog cat dog cat dog cat dog
Upper Lower Extra                                
A     a     ALPHA   3   6  -9  -6  -6  -2   0   9
            BETA   -5  -3   3  -8  -3  -2   5   8
      b     ALPHA  -4   4  -1   0   7  -4   6   6
            BETA   -9   9  -6   8   5  -2  -9  -8
B     a     ALPHA   0  -9   1  -6   2   9  -7  -9

#### 【अभ्यास】
उदाहरण के समान फ़ंक्शन को पूरा करने के लिए `rename_axis` में फ़ंक्शन का उपयोग करने का प्रयास करें, अर्थात, `Upper` और `Other` को क्रमशः `Changed_row` और `Changed_col` से बदलें।
#### 【अंत】
संपूर्ण सूचकांक के तत्वों के प्रतिस्थापन के लिए, इसे इटरेटर्स का उपयोग करके कार्यान्वित किया जा सकता है:

In [83]:
new_values = iter(list('abcdefgh'))
df_ex.rename(index=lambda x:next(new_values), level=2)

Big                 C               D            
Small               c       d       c       d    
Other             cat dog cat dog cat dog cat dog
Upper Lower Extra                                
A     a     a       3   6  -9  -6  -6  -2   0   9
            b      -5  -3   3  -8  -3  -2   5   8
      b     c      -4   4  -1   0   7  -4   6   6
            d      -9   9  -6   8   5  -2  -9  -8
B     a     e       0  -9   1  -6   2   9  -7  -9
            f      -9  -5  -4  -3  -1   8   6  -5
      b     g       0   1  -8  -8  -2   0  -6  -3
            h       2   5   9  -9   5  -6   3   1

यदि आप किसी तत्व को किसी निश्चित स्थिति में संशोधित करना चाहते हैं, तो एकल-स्तरीय सूचकांक में ऐसा करना आसान है, यानी पहले सूचकांक की `मान` विशेषता को हटा दें, फिर प्राप्त सूची को संशोधित करें, और अंत में `` को पुन: असाइन करें इंडेक्स` ऑब्जेक्ट। हालाँकि, यदि यह एक बहु-स्तरीय इंडेक्स है, तो यह थोड़ा परेशानी भरा होगा। एक समाधान यह है कि इंडेक्स के एक निश्चित स्तर को अस्थायी रूप से तालिका के एक तत्व में परिवर्तित किया जाए, फिर इसे संशोधित किया जाए और अंत में इसे रीसेट किया जाए अनुक्रमणिका। निम्नलिखित अनुभाग इन परिचालनों का परिचय देगा।एक अन्य फ़ंक्शन जिसे पेश करने की आवश्यकता है वह है `मैप`, जो `इंडेक्स` पर परिभाषित एक विधि है, यह पिछली `रीनेम` विधि में परत के कार्यात्मक उपयोग के समान है, सिवाय इसके कि यह स्केलर मान में पास नहीं होता है। परत का, लेकिन सीधे सूचकांक के टुपल में गुजरता है, जो उपयोगकर्ताओं को क्रॉस-स्तरीय संशोधन करने के लिए ट्रैवर्सल प्रदान करता है। उदाहरण के लिए, उपरोक्त स्ट्रिंग रूपांतरण ऑपरेशन को समकक्ष रूप से लिखा जा सकता है:

In [84]:
df_temp = df_ex.copy()
new_idx = df_temp.index.map(lambda x: (x[0], x[1], str.upper(x[2])))
df_temp.index = new_idx
df_temp.head()

Big                 C               D            
Small               c       d       c       d    
Other             cat dog cat dog cat dog cat dog
Upper Lower Extra                                
A     a     ALPHA   3   6  -9  -6  -6  -2   0   9
            BETA   -5  -3   3  -8  -3  -2   5   8
      b     ALPHA  -4   4  -1   0   7  -4   6   6
            BETA   -9   9  -6   8   5  -2  -9  -8
B     a     ALPHA   0  -9   1  -6   2   9  -7  -9

`मैप` का एक अन्य उपयोग बहु-स्तरीय अनुक्रमणिका को संपीड़ित करना है, जो अध्याय 4 और अध्याय 5 में कुछ कार्यों में उपयोगी है:

In [85]:
df_temp = df_ex.copy()
new_idx = df_temp.index.map(lambda x: (x[0]+'-'+x[1]+'-'+x[2]))
df_temp.index = new_idx
df_temp.head() # 单层索引

Big         C               D            
Small       c       d       c       d    
Other     cat dog cat dog cat dog cat dog
A-a-alpha   3   6  -9  -6  -6  -2   0   9
A-a-beta   -5  -3   3  -8  -3  -2   5   8
A-b-alpha  -4   4  -1   0   7  -4   6   6
A-b-beta   -9   9  -6   8   5  -2  -9  -8
B-a-alpha   0  -9   1  -6   2   9  -7  -9

साथ ही इसे विपरीत दिशा में भी विस्तारित किया जा सकता है:

In [86]:
new_idx = df_temp.index.map(lambda x:tuple(x.split('-')))
df_temp.index = new_idx
df_temp.head() # 三层索引

Big         C               D            
Small       c       d       c       d    
Other     cat dog cat dog cat dog cat dog
A a alpha   3   6  -9  -6  -6  -2   0   9
    beta   -5  -3   3  -8  -3  -2   5   8
  b alpha  -4   4  -1   0   7  -4   6   6
    beta   -9   9  -6   8   5  -2  -9  -8
B a alpha   0  -9   1  -6   2   9  -7  -9

### 3. इंडेक्स सेट करना और रीसेट करना
इस अनुभाग में कार्यों को स्पष्ट करने के लिए, आइए एक नई तालिका बनाएं:

In [87]:
df_new = pd.DataFrame({'A':list('aacd'), 'B':list('PQRT'), 'C':[1,2,3,4]})
df_new

A  B  C
0  a  P  1
1  a  Q  2
2  c  R  3
3  d  T  4

इंडेक्स सेटिंग `set_index` का उपयोग करके की जा सकती है। यहां मुख्य पैरामीटर `एपेंड` है, जो इंगित करता है कि मूल इंडेक्स को रखना है या नहीं और सीधे नई सेटिंग को मूल इंडेक्स की आंतरिक परत में जोड़ना है:

In [88]:
df_new.set_index('A')

B  C
A      
a  P  1
a  Q  2
c  R  3
d  T  4

In [89]:
df_new.set_index('A', append=True)

B  C
  A      
0 a  P  1
1 a  Q  2
2 c  R  3
3 d  T  4

आप एक ही समय में एकाधिक कॉलम को अनुक्रमणिका के रूप में निर्दिष्ट कर सकते हैं:

In [90]:
df_new.set_index(['A', 'B'])

C
A B   
a P  1
  Q  2
c R  3
d T  4

यदि आप जिस कॉलम में इंडेक्स जोड़ना चाहते हैं वह उसमें दिखाई नहीं देता है, तो आप सीधे पैरामीटर में संबंधित `सीरीज़` पास कर सकते हैं:

In [91]:
my_index = pd.Series(list('WXYZ'), name='D')
df_new = df_new.set_index(['A', my_index])
df_new

B  C
A D      
a W  P  1
  X  Q  2
c Y  R  3
d Z  T  4

`रीसेट_इंडेक्स` `सेट_इंडेक्स` का उलटा कार्य है। इसका मुख्य पैरामीटर `ड्रॉप` है, जो इंगित करता है कि हटाए गए इंडेक्स लेयर को कॉलम में जोड़ने के बजाय हटा दिया जाए:

In [92]:
df_new.reset_index(['D'])

D  B  C
A         
a  W  P  1
a  X  Q  2
c  Y  R  3
d  Z  T  4

In [93]:
df_new.reset_index(['D'], drop=True)

B  C
A      
a  P  1
a  Q  2
c  R  3
d  T  4

यदि सभी सूचकांक रीसेट हो जाते हैं, तो पांडा बस एक डिफ़ॉल्ट सूचकांक पुन: उत्पन्न कर देंगे:

In [94]:
df_new.reset_index()

A  D  B  C
0  a  W  P  1
1  a  X  Q  2
2  c  Y  R  3
3  d  Z  T  4

### 4. सूचकांक परिवर्तन
कुछ मामलों में, सूचकांक का विस्तार करना या हटाना आवश्यक है। अधिक विशेष रूप से, मूल तालिका में संबंधित सूचकांक तत्वों को नए सूचकांक से बनी तालिका में भरने के लिए एक नए सूचकांक की आवश्यकता होती है। उदाहरण के लिए, निम्न तालिका कर्मचारी जानकारी देती है एक नई तालिका बनाना, एक कर्मचारी जोड़ना, ऊंचाई कॉलम हटाना और लिंग कॉलम जोड़ना आवश्यक है:

In [95]:
df_reindex = pd.DataFrame({"Weight":[60,70,80], "Height":[176,180,179]}, index=['1001','1003','1002'])
df_reindex

Weight  Height
1001      60     176
1003      70     180
1002      80     179

In [96]:
df_reindex.reindex(index=['1001','1002','1003','1004'], columns=['Weight','Gender'])

Weight  Gender
1001    60.0     NaN
1002    80.0     NaN
1003    70.0     NaN
1004     NaN     NaN

यह आवश्यकता अक्सर समय श्रृंखला सूचकांक के समय बिंदु को भरते समय और `आईडी` संख्या का विस्तार करते समय होती है, इसके अलावा, यह ध्यान दिया जाना चाहिए कि मूल तालिका और नई तालिका में डेटा स्वचालित रूप से सूचकांक के अनुसार संरेखित हो जाएगा। उदाहरण के लिए, मूल स्थिति 1002 संख्या 1003 के बाद है, और नई तालिका इसके विपरीत है, स्थिति की परवाह किए बिना तत्वों को `रीइंडेक्स` में संरेखित किया जाएगा।

`रीइंडेक्स` के समान एक अन्य फ़ंक्शन `रीइंडेक्स_लाइक` है,जो कि पारित तालिका सूचकांक के अनुसार बुलाए गए तालिका के सूचकांक को बदलना है। उदाहरण के लिए, पहले से ही एक तालिका है जो लक्ष्य सूचकांक की शर्तों को पूरा करती है, इसलिए उपरोक्त फ़ंक्शन को निम्नलिखित कोड का उपयोग करके प्राप्त किया जा सकता है:

In [97]:
df_existed = pd.DataFrame(index=['1001','1002','1003','1004'], columns=['Weight','Gender'])
df_reindex.reindex_like(df_existed)

Weight  Gender
1001    60.0     NaN
1002    80.0     NaN
1003    70.0     NaN
1004     NaN     NaN

## 4. सूचकांक संचालन
### 1. संचालन नियम निर्धारित करें

शर्तों को पूरा करने वाली पंक्तियों को निकालने के लिए अक्सर सेट ऑपरेशंस का उपयोग करने की आवश्यकता होती है। उदाहरण के लिए, दो तालिकाएँ 'ए' और 'बी' हैं, और उनके सूचकांक कर्मचारी संख्याएँ हैं इस समय, दो तालिकाओं के अनुक्रमणिका को 'सूचकांक' पर संचालन के माध्यम से प्राप्त करना आसान है।

हालाँकि, उससे पहले, आइए चार सामान्य सेट ऑपरेशनों की समीक्षा करें:

$$\rm S_A.intersection(S_B) = \rm S_A \cap S_B \Leftrightarrow \rm \{x|x\in S_A\, और\, x\in S_B\}$$
$$\rm S_A.union(S_B) = \rm S_A \cup S_B \Leftrightarrow \rm \{x|x\in S_A\, या\, x\in S_B\}$$
$$\rm S_A.difference(S_B) = \rm S_A - S_B \Leftrightarrow \rm \{x|x\in S_A\, और\, x\notin S_B\}$$
$$\rm S_A.symmetric\_difference(S_B) = \rm S_A\triangle S_B\Leftrightarrow \rm \{x|x\in S_A\cup S_B - S_A\cap S_B\}$$

### 2. सामान्य सूचकांक संचालन

चूंकि संग्रह के तत्व अलग-अलग हैं, लेकिन सूचकांक में समान तत्व हो सकते हैं, ऑपरेशन करने से पहले डुप्लिकेट को हटाने के लिए `अद्वितीय` का उपयोग करें। निम्नलिखित प्रदर्शन के लिए दो सरलतम उदाहरण तालिकाएँ बनाता है:

In [98]:
df_set_1 = pd.DataFrame([[0,1],[1,2],[3,4]], index = pd.Index(['a','b','a'],name='id1'))
df_set_2 = pd.DataFrame([[4,5],[2,6],[7,1]], index = pd.Index(['b','b','c'],name='id2'))
id1, id2 = df_set_1.index.unique(), df_set_2.index.unique()
id1.intersection(id2)

Index(['b'], dtype='object')

In [99]:
id1.union(id2)

Index(['a', 'b', 'c'], dtype='object')

In [100]:
id1.difference(id2)

Index(['a'], dtype='object')

In [101]:
id1.symmetric_difference(id2)

Index(['a', 'c'], dtype='object')

यदि ऑपरेशन के लिए सेट किए जाने वाले दो तालिकाओं के कॉलम अनुक्रमित नहीं हैं, तो एक तरीका यह है कि पहले उन्हें इंडेक्स में परिवर्तित किया जाए, फिर ऑपरेशन के बाद उन्हें पुनर्स्थापित किया जाए। उदाहरण के लिए, `isin` फ़ंक्शन का उपयोग किया जाए पहली तालिका जहां सूचकांक रीसेट किया गया है, आईडी कॉलम के प्रतिच्छेदन वाली पंक्ति का चयन करें:

In [102]:
df_set_in_col_1 = df_set_1.reset_index()
df_set_in_col_2 = df_set_2.reset_index()
df_set_in_col_1

id1  0  1
0   a  0  1
1   b  1  2
2   a  3  4

In [103]:
df_set_in_col_2

id2  0  1
0   b  4  5
1   b  2  6
2   c  7  1

In [104]:
df_set_in_col_1[df_set_in_col_1.id1.isin(df_set_in_col_2.id2)]

id1  0  1
1   b  1  2

##5.व्यायाम
### Ex1: कंपनी कर्मचारी डेटासेट
एक कंपनी कर्मचारी डेटासेट है:

In [105]:
df = pd.read_csv('../data/company.csv')
df.head(3)

EmployeeID birthdate_key  age  city_name department      job_title gender
0        1318      1/3/1954   61  Vancouver  Executive            CEO      M
1        1319      1/3/1957   58  Vancouver  Executive      VP Stores      F
2        1320      1/2/1955   60  Vancouver  Executive  Legal Counsel      F

1. उन पुरुषों का चयन करने के लिए केवल `क्वेरी` और `लोक` का उपयोग करें जो 40 वर्ष से कम उम्र के हैं और `डेयरी` या `बेकरी` में काम करते हैं।
2. पंक्ति के पहले, तीसरे और दूसरे से अंतिम कॉलम का चयन करें जहां कर्मचारी की `आईडी` संख्या एक विषम संख्या है।
3. अनुक्रमण करने के लिए नीचे दिए गए चरणों का पालन करें:

* अंतिम तीन कॉलमों को इंडेक्स के रूप में सेट करें और आंतरिक और बाहरी परतों को स्वैप करें
* मध्य परत सूचकांक को पुनर्स्थापित करें
* बाहरी परत सूचकांक नाम को `लिंग` में बदलें
* पंक्ति अनुक्रमणिका की दो परतों को मर्ज करने के लिए अंडरस्कोर का उपयोग करें
* आरओ को विभाजित करेंw सूचकांक वापस अपनी मूल स्थिति में
* इंडेक्स नाम को मूल तालिका नाम में बदलें
* डिफ़ॉल्ट इंडेक्स को पुनर्स्थापित करें और कॉलम को मूल तालिका की सापेक्ष स्थिति में रखें

### Ex2: चॉकलेट डेटासेट
चॉकलेट समीक्षाओं के बारे में एक डेटासेट है:

In [106]:
df = pd.read_csv('../data/chocolate.csv')
df.head(3)

Company  Review\r\nDate Cocoa\r\nPercent Company\r\nLocation  Rating
0  A. Morin            2016              63%              France    3.75
1  A. Morin            2015              70%              France    2.75
2  A. Morin            2015              70%              France    3.00

1. कॉलम इंडेक्स नाम में `\n` को एक स्थान से बदलें।
2. चॉकलेट 'रेटिंग' स्कोर 1 से 5 है, 0.25 अंक अंतराल के साथ। कृपया 2.75 या उससे कम स्कोर वाले नमूने चुनें और कोको सामग्री 'कोको प्रतिशत' माध्यिका से अधिक हो।
3. `समीक्षा तिथि` और `कंपनी स्थान` को अनुक्रमणिका के रूप में सेट करने के बाद, 2012 के बाद `समीक्षा तिथि` और `कंपनी स्थान` वाले नमूने चुनें जो `फ्रांस, कनाडा, एम्स्टर्डम, बेल्जियम` से संबंधित नहीं हैं।